File: solutions/elixir/minesweeper/2/lib/minesweeper.ex
defmodule Minesweeper do
  @doc """
  Annotate empty spots next to mines with the number of mines next to them.
  """
  @spec annotate([String.t()]) :: [String.t()]

  def annotate([]), do: []
  def annotate([""]), do: [""]
  def annotate([row | _] = board) do
    tuple_board = 
      board
      |> Enum.map(&List.to_tuple(String.graphemes(&1)))
      |> List.to_tuple()

    rows = length(board)
    cols = String.length(row)

    item_accessor = fn 
      {row, col} when row in 0..rows-1 and col in 0..cols-1 -> 
        tuple_board |> elem(row) |> elem(col)
      {_, _} -> nil
    end

    accessor = fn
      {:item, row, col} -> item_accessor.({row, col})
      {:mines_count, row, col} ->
        (for i <- row-1..row+1, j <- col-1..col+1, do: item_accessor.({i, j}))
        |> Enum.filter(&(&1 == "*"))
        |> length()
    end

    for row <- 0..rows-1 do 
      for col <- 0..cols-1, reduce: "" do
        acc -> acc <> annotate(accessor, row, col)
      end
    end
  end

  defp annotate(accessor, row, col) do
    case accessor.({:item, row, col}) do
      " " -> 
        mines = accessor.({:mines_count, row, col})
        if mines > 0, do: <<?0 + mines>>, else: " "
      other -> other
    end
  end
end

File: solutions/erlang/minesweeper/1/src/minesweeper.erl
-module(minesweeper).

-export([annotate/1]).

annotate([]) -> [];
annotate([""]) -> [""];
annotate([FirstRow | _] = Board) ->
    TupleBoard = list_to_tuple(lists:map(fun (Row) -> list_to_tuple(Row) end, Board)),
    Rows = length(Board),
    Cols = length(FirstRow),
    BoardAccessor = fun 
        ({X, Y}) when X >= 0 andalso X < Rows andalso Y >= 0 andalso Y < Cols ->
            element(Y+1, element(X+1, TupleBoard));
        ({_, _}) -> nil
    end,
    Accessor = fun
        ({item, X, Y}) -> BoardAccessor({X, Y});
        ({mines_count, X, Y}) -> 
            Cells = [BoardAccessor({I, J}) || I <- lists:seq(X - 1, X + 1), J <- lists:seq(Y - 1, Y + 1)],
            length(lists:filter(fun(Cell) -> Cell == $* end, Cells))
    end,
    lists:map(
        fun (X) -> [annotate_cell(Accessor, X, Y) || Y <- lists:seq(0, Cols-1)] end,
        lists:seq(0, Rows-1)).

annotate_cell(Accessor, X, Y) ->
    case Accessor({item, X, Y}) of
        $\s ->
            Mines = Accessor({mines_count, X, Y}),
            if 
                Mines > 0 -> $0 + Mines;
                true -> $\s
            end; 
        Other -> Other
    end.
